File: 0865_Smallest_Subtree_with_all_the_Deepest_Nodes/solution.cpp
/*
865. Smallest Subtree with all the Deepest Nodes
https://leetcode.com/problems/smallest-subtree-with-all-the-deepest-nodes/
 */

#include <algorithm>
using namespace std;

#include "common/TreeNode.h"
/*
https://zxi.mytechroad.com/blog/tree/leetcode-865-smallest-subtree-with-all-the-deepest-nodes/

 */

class Solution
{
public:
  TreeNode *subtreeWithAllDeepest(TreeNode *root)
  {
    return depth(root).second;
  }

private:
  pair<int, TreeNode *> depth(TreeNode *root)
  {
    if (root == nullptr)
      return {-1, nullptr};
    auto l = depth(root->left);
    auto r = depth(root->right);
    int dl = l.first;
    int dr = r.first;
    return {max(dl, dr) + 1,
            dl == dr ? root : (dl > dr) ? l.second : r.second};
  }
};

class Solution
{
public:
  TreeNode *subtreeWithAllDeepest(TreeNode *root)
  {
    TreeNode *ans;
    depth(root, &ans);
    return ans;
  }

private:
  int depth(TreeNode *root, TreeNode **s)
  {
    if (root == nullptr)
      return -1;
    TreeNode *sl;
    TreeNode *sr;
    int l = depth(root->left, &sl);
    int r = depth(root->right, &sr);
    *s = (l == r) ? root : ((l > r) ? sl : sr);
    return max(l, r) + 1;
  }
};

File: 0865_Smallest_Subtree_with_all_the_Deepest_Nodes/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.TreeNode;

/*
865. Smallest Subtree with all the Deepest Nodes

Given a binary tree rooted at root, the depth of each node is the shortest distance to the root.

A node is deepest if it has the largest depth possible among any node in the entire tree.

The subtree of a node is that node, plus the set of all descendants of that node.

Return the node with the largest depth such that it contains all the deepest nodes in its subtree.
 */
public class LC_0865_SmallestSubtreewithalltheDeepestNodes {
    class Result {
        int depth;
        TreeNode node;

        public Result(TreeNode node, int depth) {
            this.depth = depth;
            this.node = node;
        }
    }
    /*
       Case 1: depth left = depth right, return root
       Case 1: depth left > depth right, return answer of root->left
     */
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        TreeNode result = null;
        return depth(root).node;
    }

    private Result depth(TreeNode root) {
        if (root == null)
            return new Result(null, 0);
        Result left = depth(root.left);
        Result right = depth(root.right);
        if (left.depth == right.depth)
            return new Result(root, left.depth + 1);
        else if (left.depth > right.depth)
            return new Result(left.node, left.depth + 1);
        else
            return new Result(right.node, right.depth + 1);
    }
}
